### PR TITLE
5434 - File upload error reporting fix for large files

### PR DESCRIPTION
**Pull request by qqmyers**
which broke error reporting. This update should now add error reporting
to provenance uploads as well.

## New Contributors

Welcome! New contributors should at least glance at [CONTRIBUTING.md](/CONTRIBUTING.md), especially the section on pull requests where we encourage you to reach out to other developers before you start coding. Also, please note that we measure code coverage and prefer you write unit tests. Pull requests can still be reviewed without tests or completion of the checklist outlined below. Thanks!

## Related Issues

- connects to #5471 Broken error checking 

Relates to #5434 File upload: large file uploads should show up on file page after each ingest, before save

## Pull Request Checklist

- [ ] Unit [tests][] completed
- [ ] Integration [tests][]: None
- [ ] Deployment requirements, [SQL updates][], [Solr updates][], etc.: None
- [ ] [Documentation][docs] completed
- [ ] Merged latest from "develop" [branch][] and resolved conflicts

[tests]: http://guides.dataverse.org/en/latest/developers/testing.html
[SQL updates]: http://guides.dataverse.org/en/latest/developers/sql-upgrade-scripts.html
[Solr updates]: https://github.com/IQSS/dataverse/blob/develop/conf/solr/7.3.0/schema.xml
[docs]: http://guides.dataverse.org/en/latest/developers/documentation.html
[branch]: http://guides.dataverse.org/en/latest/developers/branching-strategy.html

### CODE REVIEW

File: src/main/webapp/provenance-popups-fragment.xhtml [hunk 1]
```diff
@@ -48,19 +49,11 @@
                                           fileUploadListener="#{provPopupFragmentBean.handleFileUpload}" 
                                           update="provPopupOutput,assignMessages" 
                                           label="#{bundle['file.editProvenanceDialog.selectToAddBtn']}"
+                                          onstart="javascript:uploadWidgetDropRemoveMsg();uploadStarted();"
+                                          onerror="javascript:uploadFailure();"
                                           oncomplete="bind_bsui_components();PF('provUpload').hide();"
                                           fileLimit="1" 
                                           invalidSizeMessage="#{bundle['file.edit.error.file_exceeds_limit']}"/>
-                            <script>
-                                function uploadStarted() {
```
Comment (qqmyers) [inline src/main/webapp/provenance-popups-fragment.xhtml]: FWIW - this uploadStarted() method was overwriting the one in editFilesFragment.xhtml...

File: src/main/webapp/editFilesFragment.xhtml [hunk 2]
```diff
@@ -77,113 +78,12 @@
                                 </p>
                     
                     <script>
-//<![CDATA[
```
Comment (mheppler) [inline src/main/webapp/editFilesFragment.xhtml]: Jim, any particular reason to remove this `CDATA` section from the script?

Comment (qqmyers) [reply src/main/webapp/editFilesFragment.xhtml]: I think it was added when there were parse errors due to the content of the javascript. With most of that now in a separate file, I didn't think it was needed.

**Comment by mheppler** [comment]
@qqmyers Could you please create a new issue, specific to the error logging vs provenance `uploadFinished()` fix, to connect this PR to, rather than #5434 which may be a larger issue.

Also, I have asked @matthew-a-dunlap to take a look at the provenance code to make sure he's OK with the changes.

**Comment by qqmyers** [comment]
@mheppler - new issue and edited the PR to point at it - let me know if that's not what you wanted.

**Comment by matthew-a-dunlap** [comment]
I have this branch running to test, specifically related to the provenance changes. TBH I'm not 100% sure of the need for that provenance js code in the first place.

That being said, I'm running into the attached error trying to do normal file uploads via http. Develop branch is working for me fine.

[uploaderror.log](https://github.com/IQSS/dataverse/files/2766200/uploaderror.log)

**Comment by qqmyers** [comment]
@matthew-a-dunlap - as far as I can tell, the only thing this PR is doing is showing an existing (perhaps relatively new) issue: The error code is is getting triggered by PrimeFace uploading when it calls the onerror method of the p:fileUpload element. The HTTP status I see is 200, meaning the call went through fine (versus 502 errors for example). Without the javascript error code, this would be a silent issue on the browser side (so it could be occurring on develop and only showing in the log?)

The duplicate element appears to be related to dataset linking. I haven't followed what changes have been in that area lately, but I wonder if some change there is causing an issue when the page is being refreshed during upload. In testing last night, I only saw the error code show a 200 error, with similar log entries as you show, when I uploaded files in multiple actions (e.g. multiple hits on 'the select files to upload' button). I think I saw silent failures like this when I was working on issue #5459, before I repaired the error code, which would be consistent with it being an existing issue (again perhaps a recent one).

It's possible that there's some interaction with the javascript involved but I'm not sure how. The only code called before the error is returned is the uploadStarted(), which adds a 'upid' attribute to some nodes and starts mutators (event watchers) but doesn't otherwise change/add nodes (and it definitely doesn't touch the data linking code which is in the underlying dataset.xhmtl file rather than the editFilesFragment.xhtml. I don't think any code in this PR changes what parts of the page/form are refreshed, etc.

If the log entries are showing up in the develop branch, we should create a separate issue. If not, I can try removing parts of the uploadStarted() method to see what is triggering the issue. 

For the longer term, I could also change the error handler to not report status 200 errors, or to highlight them - thinking beyond the current issue, I think they will always mean a bug/something dataverse could/should handle, versus problems caused by proxies, etc. that are config/network issues, so they could be hidden (let the log report the problem) or flagged differently to ask users to report them to dataverse, etc.)

**Comment by matthew-a-dunlap** [comment]
@qqmyers - Ahh I missed that it was a 200. My experience with this branch is that on the create datatset page the file uploads are now broken. I've had it succeed rarely, most of the time it throws the red error message with a 200 status code. Saving after that the file is not part of the dataset. Uploading after dataset creation seems to work every time. 

With the log messages, the duplicate component ID message may be a red herring. I think I'm mostly worried about the large javax.enterprise message hitting our logs at severe level. If that 750kbs message is intended we should probably set it to a lower level and only turn it on when we need to debug. Maybe this was in here before the code move-around but we don't want to fill up our logs too much.

**Comment by qqmyers** [comment]
@matthew-a-dunlap - I think I found the issue: my change restored functionality that was evidently obsolete. The original uploadStarted method in editFilesFragment clicks the uploadStarted button, but the method in the prov popup that replaced it did not. Removing that (latest commit) appears to fix the error in create mode. I'll leave the error code where it will show a 200 error if one happens (this has no effect on what's logged, so it won't affect log size if some other 200 error shows up.)

**Comment by matthew-a-dunlap** [comment]
@qqmyers Interesting! The prov popup code was not intended to override this functionality. Any idea why we didn't run into upload issues before adding the popup?

**Comment by qqmyers** [comment]
@matthew-a-dunlap - not really, other than the possible hint that the duplicate item is in data linking and I thought (?) there were updates to that code after provenance was merged. If I"m following the code correctly that button only set uploadInProgress=true which is set true by the upload handlers anyway, so it's not clear to me that the call was needed. The button has update="", so I also don't really understand why making the call causes duplicate elements in the first place unless "" doesn't mean don't update anything.

**Comment by mheppler** [comment]
@qqmyers @matthew-a-dunlap Since the linking popup on the dataset was resulting in the duplicate ID errors, I looked in there for what changed recently (see #4410).

When I started searching for causes of `Duplicate component ID` errors, I found a reference in this [PrimeFaces GitHub issue](https://github.com/primefaces/primefaces/issues/1691) which links to this [Stack Overflow question](https://stackoverflow.com/questions/2101755/binding-attribute-causes-duplicate-component-id-found-in-the-view) about "binding attribute causes duplicate component ID".

It looks like the new dataset linking popup code includes a `p:autoComplete` component with a `binding="#{DatasetPage.selectedDataverseMenu}"` attribute. I don't know enough about what it is used for, and if this is the cause of the excessive error logging, but it might be worth looking into.

To be fair, it looks like we use the `binding` attribute all over the app. This is the only instance of it on the `dataset.xhtml` pg though.

File: src/main/webapp/editFilesFragment.xhtml [hunk 3]
```diff
@@ -377,7 +276,6 @@
                     
                     <p:message for="fileUpload" id="uploadMessage" display="text" />
                     
-                    <p:commandButton id="uploadStarted" action="#{EditDatafilesPage.uploadStarted()}" update="" style="display:none"/>
```
Comment (landreev) [inline src/main/webapp/editFilesFragment.xhtml]: This was part of a legacy fix/hack... But it looks like we are not using it anymore for anything useful... So should be ok to remove as well.

File: src/main/webapp/provenance-popups-fragment.xhtml (hunk 1 above)
Comment (landreev) [reply src/main/webapp/provenance-popups-fragment.xhtml]: @matthew-a-dunlap Could you confirm that this method was not needed, for the whole provenance upload thing to work? (looks like not - I'm not even seeing a button with the id "provUploadStarted"... but wanted to double-check)

Comment (matthew-a-dunlap) [reply src/main/webapp/provenance-popups-fragment.xhtml]: @landreev I am pretty sure it doesn't, I'm hoping to check in on it while waiting for my flight tomorrow.

**Comment by matthew-a-dunlap** [comment]
I'm away on vacation / a conference this coming week. I may have time to dig into this more but it is unsure. If someone else has a chance to dig in, provenance can be enabled via command line http://guides.dataverse.org/en/latest/installation/config.html#provcollectionenabled

**Comment by qqmyers** [comment]
@matthew-a-dunlap  - I've fixed the first error - the prov on start call shouldn't call uploadWidgetDropRemoveMsg (which deletes the 'drag and drop files here message on the upload files fragment). The second error -also fixed, is not specific to this branch. The prov oncomplete method is trying to hide the provUpload div (which has the select files button on the prov fragment) but it is deleted/replaced in the DOM by the 'upload complete'/preview/remove elements. So- hiding is not needed.

With these commits, I think everything found so far has been addressed.

**Comment by matthew-a-dunlap** [comment]
@qqmyers This looks great! I tested many of the permutations of the prov functionality across dataset-create / file / edit-metadata and the all look good. Thanks for cleaning up those FileUpload components, this was my first major piece of jsf work and pieces got left behind as I was reusing code.

**Pull request merged**